### ISSUE TITLE
Concur connector and non-required parameters

### ISSUE THREAD

**Issue by mcada** [syndesisio/syndesis#3377]
## This is a...

<pre><code>
[ ] Feature request
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Documentation issue or request
</code></pre>

## The problem

Integration with concur connection fails with exception if you do not specify all parameters required by selected concur action. But for example action "Gets all lists" we can see here https://developer.concur.com/api-explorer/v3-0/Lists.html that those fields are not required and have default values but it is still required from syndesis and does not work without them: 

```
java.lang.IllegalArgumentException: MIME type may not contain reserved characters
	at org.apache.http.util.Args.check(Args.java:36) ~[httpcore-4.4.9.jar!/:4.4.9]
	at org.apache.http.entity.ContentType.create(ContentType.java:227) ~[httpcore-4.4.9.jar!/:4.4.9]
	at org.apache.http.entity.ContentType.create(ContentType.java:239) ~[httpcore-4.4.9.jar!/:4.4.9]
	at org.apache.camel.component.http4.HttpProducer.createRequestEntity(HttpProducer.java:511) ~[camel-http4-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.component.http4.HttpProducer.createMethod(HttpProducer.java:455) ~[camel-http4-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.component.http4.HttpProducer.process(HttpProducer.java:113) ~[camel-http4-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.util.AsyncProcessorConverterHelper$ProcessorToAsyncProcessorBridge.process(AsyncProcessorConverterHelper.java:61) ~[camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.component.rest.RestProducer.process(RestProducer.java:86) ~[camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:138) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:101) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.component.connector.ConnectorProducer.process(ConnectorProducer.java:45) ~[camel-connector-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.TryProcessor.process(TryProcessor.java:113) ~[camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.TryProcessor.process(TryProcessor.java:84) ~[camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:138) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:101) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.component.connector.ConnectorProducer.process(ConnectorProducer.java:45) ~[camel-connector-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.SendProcessor.process(SendProcessor.java:148) ~[camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.DelegateAsyncProcessor.process(DelegateAsyncProcessor.java:110) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.RedeliveryErrorHandler.process(RedeliveryErrorHandler.java:548) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:201) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:138) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:101) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.DelegateAsyncProcessor.process(DelegateAsyncProcessor.java:110) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at io.syndesis.integration.runtime.logging.ActivityTrackingInterceptStrategy$EventProcessor.process(ActivityTrackingInterceptStrategy.java:79) [integration-runtime-1.4.5.jar!/:1.4.5]
	at org.apache.camel.processor.DelegateAsyncProcessor.process(DelegateAsyncProcessor.java:110) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.RedeliveryErrorHandler.process(RedeliveryErrorHandler.java:548) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:201) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:138) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:101) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.RedeliveryErrorHandler.process(RedeliveryErrorHandler.java:548) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:201) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.MulticastProcessor.doProcessSequential(MulticastProcessor.java:715) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.MulticastProcessor.doProcessSequential(MulticastProcessor.java:638) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.MulticastProcessor.process(MulticastProcessor.java:248) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.Splitter.process(Splitter.java:122) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.DelegateAsyncProcessor.process(DelegateAsyncProcessor.java:110) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.RedeliveryErrorHandler.process(RedeliveryErrorHandler.java:548) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:201) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:138) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:101) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:201) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.component.timer.TimerConsumer.sendTimerExchange(TimerConsumer.java:197) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at org.apache.camel.component.timer.TimerConsumer$1.run(TimerConsumer.java:79) [camel-core-2.21.0.fuse-000094.jar!/:2.21.0.fuse-000094]
	at java.util.TimerThread.mainLoop(Timer.java:555) [na:1.8.0_151]
	at java.util.TimerThread.run(Timer.java:505) [na:1.8.0_151]
```
## Expected behavior

Optional API parameters should not be required from syndesis. 

## Tasks involved / Steps to Reproduce

1. Create integration with concur action "Gets all lists"
2. See that it requires data mapper even though those parameters are optional in concur API

**Comment by zregvart** [comment]
The exception is noting that the `Content-Type` Camel message header is missing on the exchange. Could be the side effect of not specifying any parameters and the body being `null` or similar.

@mcada can you provide an outline of the integration, is there any mapping used? Perhaps you could add a log step before the Concur connection and see what is contained in the message body and headers when there is a parameter specified and when there isn't a parameter specified to narrow this?

**Comment by mcada** [comment]
**Integration:**
Database -> log -> concur (Get all lists) -> log

Error:
```
ERROR 1 --- [r://integration] o.a.camel.processor.DefaultErrorHandler  : Failed delivery for (MessageId: i-LJsPH3zRAxPXvUf5vsQz on ExchangeId: i-LJsPH1HRAxPXvUf5vsOz). Exhausted after delivery attempt: 1 caught: java.lang.IllegalArgumentException: MIME type may not contain reserved characters
```
Outcome:
There was notification that data mapper is required, yet both fields are optional in concur API. This should work. 

**Integration:**
Database -> log -> mapper (no mappings) -> concur (Get all lists) -> log

Error:
```
ERROR 1 --- [r://integration] o.a.camel.processor.DefaultErrorHandler  : Failed delivery for (MessageId: i-LJsRUSF7RbgBhf4Zn4bz on ExchangeId: i-LJsRUOj7RbgBhf4Zn4_z). Exhausted after delivery attempt: 1 caught: java.io.FileNotFoundException: Cannot find resource: mapping-step-3.json in classpath for URI: mapping-step-3.json
```
Outcome:
No mappings in data mapper, I expected this to fail anyway.

**Integration:**
Database -> log -> mapper (one mapping field mapped) -> concur (Get all lists) -> log

Error:
```
ERROR 1 --- [r://integration] i.s.c.r.s.OAuthRefreshTokenProcessor     : Unable to refresh the access token, received status: `HTTP/1.1 400 Bad Request`, response: `{"error":"invalid_request","error_description":"client_id is not known to us","concur-correlationid":"a10ece76-6450-471e-9dfb-fe5f3da82dba","geolocation":"https://us.api.concursolutions.com","code":61}`
```
For some reason, I lost here OAuth credentials in settings page. I only edited previous integration and added one mapping step and clicked publish.

Filled credentials again, stopped integration. Syndesis error popped out:
```Error performing GET request to https://syndesis.my-minishift.syndesis.io/api/v1/integrations?per_page=50 : ```
```status: 502, statusText: "Bad Gateway```
another error incomming
```status: 504, statusText: "Gateway Time-out```

note that I work locally on minishift, I have never experienced such errors before.
After some time, integrations page loaded, integration was still running, so clicked again stop, confirmed stop.
Nothing happened, ctrl+f5 and I see that integration is indeed stopped. Starting updated integration. Again the same errors with gateway... Going to reinstall syndesis...

After reinstalling syndesis, integration started. Note that I use syndesis 1.4.5 so there was no pod update. I worked on this syndesis instance just one day and it failed like this. Reinstall helped.

Outcome: I expected this to fail, but this integration worked so setting at least one optional field worked.




**Integration:** - different concur action
Database -> log -> mapper (mapped one filed) -> concur (Get all listitems) -> log

Outcome:
Again I mapped at least one field otherwise it fails as we already know. It worked, I mapped field "limit" from newly created int constant with value 50. 

**Integration:** - yet another concur action
Database -> log -> mapper (mapped one filed) -> concur (create a new listitem) -> log
Note: I mapped all required fields (Name, ListId, Level1Code) - this worked for me straight from postman to concur without any issues

Error:
```
ERROR 1 --- [r://integration] o.a.camel.processor.DefaultErrorHandler  : Failed delivery for (MessageId: i-LJs_qyfYc2HSgZ85Xhdz on ExchangeId: i-LJs_quEYc2HSgZ85Xh_z). Exhausted after delivery attempt: 1 caught: org.apache.camel.http.common.HttpOperationFailedException: HTTP operation failed invoking https://implementation.concursolutions.com/api/v3.0/common/listitems with statusCode: 400
```

Outcome:
I expected this to pass, but something is wrong here and I do not know what. From logs, I am not sure what should I do differently. I used correct ListID, all other fields can have new value and it worked via postman, cannot do it via integration. Maybe @KurtStam knows what is wrong here?
From logs: 
```listId  has an invalid value. Please Check the API documentation to learn how to provide valid values.```
But I took listId from previous integration output (get all lists) and selected existing list and took its listId. I did the same with postman and listId was the same and it worked there. Could it be problem with data mapper and special characters like "$"? I don't think so, but for some reason it does not work. 

As long as I am not able to create concur listItems from integration, our tests will not work.

[concurCreateListitemError.txt](https://github.com/syndesisio/syndesis/files/2287135/concurCreateListitemError.txt)

**Comment by mcada** [comment]
TLDR of previous comment: even if concur API endpoint has no required parameter, syndesis requires at least one parameter set otherwise it does not work. 
Also I am not able to create listItem from integration even though through postman it works with the same values.

**Comment by KurtStam** [comment]
@zregvart do you think the api-connector in general is not sending a proper message when no values are set?

**Comment by zregvart** [comment]
@KurtStam I think when no parameters are set we don't generate a Camel message body and the Camel message header that contains the `Content-Type` header.

We could set the `Content-Type` header here:
https://github.com/syndesisio/syndesis/blob/0e51030367eaac7edb1a7c6dd5d5247c2a565bc6/app/connector/rest-swagger/src/main/java/io/syndesis/connector/rest/swagger/SwaggerConnectorComponent.java#L198-L219

based on the OpenAPI specification. That is determine one, if it's not already present on the incoming Camel message, by looking at the `consumes` MIME types of the OpenAPI operation that we're trying to invoke.

Could also be related to #3388

**Comment by KurtStam** [comment]
I can no longer reproduce this. It seems to work ok with no parameters specified:

<img width="928" alt="screen shot 2018-08-28 at 9 44 01 am" src="https://user-images.githubusercontent.com/35576/44727049-185fec00-aaa7-11e8-9597-db97e5e6170b.png">

**Comment by KurtStam** [comment]
Please reopen if this is still an issue.
